### ISSUE TITLE
Curry-safe reduce?

### ISSUE THREAD

**Issue by rjmk** [sanctuary-js/sanctuary#163]
Is there space in sanctuary for a reduce-like function that calls the reducing function once on the accumulator and then the result on the current element? As it stands, Sanctuary, like in Ramda, doesn't work for manually curried functions.

**Comment by davidchambers** [comment]
I like this proposal. Having higher-order functions operate on non-unary functions makes me uneasy. I'm open to the possibility of supporting both. For example:

```haskell
reduce  :: Foldable f => (a -> b -> a) -> a -> f b -> a
reduce_ :: Foldable f => ((a, b) -> a) -> a -> f b -> a
```

`reduce` would support manually curried functions, `reduce_` would support true binary functions, and both would support functions curried Ramda-style.

If there's support for this approach I suggest we apply it throughout the library.

**Comment by rjmk** [comment]
> If there's support for this approach I suggest we apply it throughout the library.

I would really like to see this! When I'm working, it more often occurs to me to use sanctuary so I can declutter and write a better pointfree definition than for the type-safety (not that I care more about pointfree than type-safety, it's just more salient)

**Comment by davidchambers** [comment]
Here are the higher-order functions in question:

```console
$ pcregrep '//# .* :: .*[(][^()]* -> [^()]* -> [^()]*[)]' index.js
  //# C :: (a -> b -> c) -> b -> a -> c
  //# S :: (a -> b -> c) -> (a -> b) -> a -> c
  //# flip :: (a -> b -> c) -> b -> a -> c
  //# lift2 :: Apply f => (a -> b -> c) -> f a -> f b -> f c
  //# lift3 :: Apply f => (a -> b -> c -> d) -> f a -> f b -> f c -> f d
  //# meld :: [** -> *] -> (* -> * -> ... -> *)
  //# Maybe#reduce :: Maybe a ~> (b -> a -> b) -> b -> b
  //# encase2 :: (a -> b -> c) -> a -> b -> Maybe c
  //# encase3 :: (a -> b -> c -> d) -> a -> b -> c -> Maybe d
  //# encaseEither2 :: (Error -> l) -> (a -> b -> r) -> a -> b -> Either l r
  //# encaseEither3 :: (Error -> l) -> (a -> b -> c -> r) -> a -> b -> c -> Either l r
  //# reduce :: Foldable f => (a -> b -> a) -> a -> f b -> a
```

**Comment by rjmk** [comment]
The combinators (`C`, `S`) and the `lift`s are already appropriately defined, so that only leaves the other 8

**Comment by davidchambers** [comment]
Let's change the representation of `flip`'s first argument's type from `(a -> b -> c)` to `((a, b) -> c)` while we're at it.

**Comment by davidchambers** [comment]
Resolved by #182
